[WIP] user_settings: Add push notifications for all stream messages.

sarahcstringer opened the pull request:
Add setting to enable push notifications for all stream messages.

Checking in to see if this is the correct direction for this setting. 

Also, 2 tests are failing because of the increase in the number of queries running:
- test_not_too_many_queries (zerver.tests.test_messages.StreamMessagesTest): goes from 16 to 22 queries.
- test_multi_user_subscription (zerver.tests.test_subs.SubscriptionAPITest): goes from 43 to 45 queries.

The extra queries are happening in `zerver/lib/actions` in the `do_send_messages` function. I added a new function, `subscribed_to_stream_push_notifications`, to check if the user is subscribed to get push notifications for a particular stream. I'm wondering if the extra queries are too much and I should rework the code to do fewer, or if I should just update the tests?

I have not yet added tests for the new functionality; that's next on my list.

Addresses issue #6132

timabbott commented on static/js/stream_data.js:
```diff
@@ -392,6 +393,14 @@ exports.receives_audible_notifications = function (stream_name) {
+        return false;
+    }
+    return sub.push_notifications;
+};
```
This doesn't seem to be called?  But just as a note, we're trying to migrate away from accessing objects by `stream_name` wherever we can use `stream_id` instead.

timabbott commented on static/templates/subscription_settings.handlebars:
```diff
@@ -61,6 +61,14 @@
+                            <input id="push-notifystream-{{stream_id}}" class="sub_setting_control" type="checkbox" tabindex="-1" {{#if push_notifications}}checked{{/if}} {{#unless in_home_view}}disabled="disabled"{{/unless}}/>
+                            <label class="subscription-control-label">{{t "Receive push notifications" }}</label>
+                        </div>
+
```
I think this newline was a bug in the previous one that shouldn't be copied :)

timabbott commented on static/templates/subscription_settings.handlebars:
```diff
@@ -61,6 +61,14 @@
                     <li>
+                        <div id="sub_push_notifications_setting"
+                            class="sub_setting_checkbox sub_notification_setting {{#unless in_home_view}}muted-sub{{/unless}}">
+                            <input id="push-notifystream-{{stream_id}}" class="sub_setting_control" type="checkbox" tabindex="-1" {{#if push_notifications}}checked{{/if}} {{#unless in_home_view}}disabled="disabled"{{/unless}}/>
```
We might not need some of the semi-duplicate IDs here?  An independent cleanup to consider while we're at it, at least.  Could make sense as a quick commit on the bottom of the branch.

sarahcstringer replied:
Was it the div ids that were unnecessary, and instead keep the input ids?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -897,7 +897,9 @@ def do_send_messages(messages_maybe_none):
                   'flags': user_flags.get(user.id, []),
-                  'always_push_notify': user.enable_online_push_notifications}
+                  'always_push_notify': user.enable_online_push_notifications,
+                  'stream_push_notify': subscribed_to_stream_push_notifications(user, message['stream']),
```
So the crux of the performance problem is this call; we're doing 1 database query per user subscribed to the stream, which is exactly what the test is meant to prevent us from doing, since it means if you send to a stream with 2000 subscribers, we're going to do 2000 database queries.

What you want to do instead is patch `get_recipient_user_profiles`, which is already querying all the relevant Subscription objects.  It should add to the `fields` list for streams the field `push_notifcations`, and then do the somewhat annoying logistics to pass the data back to the callers of `get_recipient_user_profiles`.  Maybe the right logistics is to return a dict `sub_dict` mapping {user_profile_id: sub_object}, and then the code here can access `sub_dict[user.id]['push_notifications']`.  It'll be a bit messier than that, because `sub_dict` will be empty for non-stream messages.

Can you try playing with that approach?  I'd definitely rebase onto master before doing more here because we just merged a PR that refactors `recipient_user_profiles` or its neighbors; can't remember exactly.

timabbott commented on static/templates/settings/notification-settings.handlebars:
```diff
@@ -36,6 +36,20 @@
+                    <span></span>
+                </label>
+                <label for="enable_stream_push_notifications" class="inline-block no-border">
+                    {{t "Push notifications" }}
```
Maybe call this "Mobile push notifications", for clarity?

zulipbot-test commented:
Heads up @sarahcstringer, we just merged some commits that conflict with the changes your made in this pull request! You can review this repository's [recent commits](https://github.com/zulip/zulip/commits/master) to see where the conflicts occur. Please rebase your feature branch against the `upstream/master` branch and [resolve](http://zulip.readthedocs.io/en/latest/git-guide.html#recover-from-a-git-rebase-failure) your pull request's merge conflicts accordingly.

sarahcstringer commented:
Thank you for your help @timabbott. I think I fixed the issue with the multiple queries. I'm still working on tests for the new feature.

sarahcstringer commented:
Apologies for the delay on this. I just pushed some additional changes. Updates:
- Added tests for the settings I added, but I think there are still tests that I need to write to see if stream messages actually trigger the correct push notifications -- trying to figure that one out :) 
- It looks like there are a few places that use the ids on the semi-duplicate divs that Tim mentioned, so I didn't make that suggested change yet (there are tests and event listeners using the div ids).

sarahcstringer commented:
I updated the alerts portion of push_notifications because now if someone receives a push notification from a stream message, it can either be because they were mentioned or because they're electing to get all stream messages as push notifications. The previous alert for stream messages assumed they were mentions.

timabbott commented:
There's some significant merge conflicts in `actions.py` with some changes @showell made to optimize performance.  I suspect it should be straightforward to migrate it, but it might make sense for @showell to take a quick look at whether there's anything special we need to do to avoid regressing perf (or just doing the rebase).  We do want this feature in Zulip 1.7 if we can get it in.

showell commented:
I'm working through the rebase now.

showell commented:
see #6484

timabbott commented:
OK, I gave this a thorough review and merged most of it; notes are below.  

I merged the first commit of #6484.  

For the second commit, everything looks good except that I think we don't want to change the `idle or always_push_notify` block; I believe the intent of this feature is have stream messages on tagged streams be treated like PMs, not always push notify.  So we should follow that convention.  I merged it with that change (well, I also bumped PROVISION_VERSION).  

For the third commit, it looks good, except that the `received_pm` name for that trigger is kinda a bad variable name to be now exposing to the mobile apps via an API, since we don't use abbreviations, and also every case is `received`, so it's not similar to the others.  I renamed it to `private_message` in a new commit just after this one.  Also, I reverted the small `logging.info` text change, since we generally want to avoid putting user data into log files, and that could include stream names (etc.).

The last commit of this PR doesn't do the i18n correctly.  I think there are 2 issues:
(1) The string should be tagged for translation before variable substitution, otherwise it will never match.  See Django's i18n docs for how to do this.  
(2) I think we might need to use `with_language` from zerver/lib/i18n to make sure that we're actually using the language of the recipient of the push notification, not the language of the person sending the message that triggered it.

A few follow-ups from the things that are merged:
* We should track down why this added a new database query to sending messages to streams.  I don't see why it should have, and every query in that code path counts.  @sarahcstringer can investigate this?  @showell may be able to help debug, but you can get Zulip to log every database query by uncommenting the `django.db` logging section in `zproject/settings.py`, which may be enough to find out what query is new here. 
* For the push notification tests, I think we probably want to, rather than hardcoding values for the `triggers` section, instead see if we can call a function that will compute them automatically from the draft message.  May involve some refactoring, but could cut out a lot of test code duplication.
* @rishig the text for the notification is currently "New stream message from %s in Denmark".  What should it be?  Might want to use a `#` before the stream name, and maybe we don't need the sender?  Relatedly, we currently have ` "New mention from %s" % (sender_str,)`; not sure if that should highlight the stream name.
* We should update the user documentation for this.  @synicalsyntax can help potentially help with this, but basically just look at the stuff under `templates/server/help` for the pre-existing similar features and following the pattern should work. 

@borisyankov @kunall17 I'm deploying this to chat.zulip.org shortly.  We should make sure the feature works end-to-end; it might require some changes in the mobile app's formatting code depending whether you're doing your own formatting or using the text provided by the server.  To test, just set a the "test here" stream to send push notifications in the stream settings UI, and test push notifications as usual.  One particular bug to look for is whether we trigger pushes for messages you yourself send to a stream; we probably don't want to do that, and the code looks correct, but worth double-checking end-to-end).  

We may also want to open an zulip-mobile issue for being able to toggle this setting from the app, since that seems like a useful interface.

Huge thanks for making this happen @sarahcstringer!

timabbott commented:
@sarahcstringer oh, I guess we'd lost track of the last commit on this PR; maybe it's worth extracting that to a new PR?  This one has a lot of history on it.
